@std/fs: replace mkdir with createdirectory to handle pathlike objects and create intermediate directories

annieetang opened the pull request:
### Problem

Resolves https://github.com/luau-lang/lute/issues/413

We're currently using `fs.mkdir` for all use cases without the option of making intermediate parent folders, and the `mkdirdashp` function in lute setup commands can be moved to the `std/fs` library for broader use cases.

### Solution

We replace `fs.mkdir` in the standard library with `fs.createdirectory` (a more user friendly name!) that takes in a `path: pathlike` argument as well as an `options` type to specify if the user wants to make intermediate parent folders with `makeparents: boolean` 
For example: `fs.createdirectory(tmpdir, { makeparents = true })` (see more in `examples/create_directory.luau`)

We also update the lute setup commands file to use `fs.createdirectory()` and add/update relevant tests

annieetang committed:
std/fs replace mkdir with createdirectory(pathlike, options)

Vighnesh-V commented on lute/std/libs/fs.luau:
```diff
@@ -88,6 +91,26 @@ local function readasync(filepath: pathlike): string
 	return fs.readasync(pathlib.format(filepath))
 end
 
+local function createdirectory(path: pathlike, options: createdirectoryoptions): ()
```
These options should be optional! When they're nil, this should just do mkdir under the hood.

annieetang committed:
Merge branch 'primary' of https://github.com/luau-lang/lute into annietang/fs_createdirectory

annieetang committed:
update fs.test.luau

annieetang committed:
replace mkdirdashp use with fs.createdirectory

annieetang committed:
cleanup

annieetang committed:
fix string and pathlike concats in cli setup init.luau

annieetang committed:
use pathlike types for luaurcpath processing too

annieetang committed:
add example in examples/

aatxe commented on examples/create_directory.luau:
```diff
@@ -0,0 +1,5 @@
+local fs = require("@std/fs")
+local path = require("@std/path")
+
+local directory = path.parse("example/new_directory/subdir")
```
maybe make this make the directory in tmpdir?

annieetang replied:
ah this was just an example folder but I can pop it in a tmpdir

aatxe commented on lute/cli/commands/setup/init.luau:
```diff
@@ -50,11 +36,12 @@ if not table.find(args, "--with-luaurc") then
 end
 
 local luauRcPath = `{cwd}/.luaurc`
+local luauRcPathPathlike = path.parse(luauRcPath)
```
do paths not have an `__tostring`? if they do (which they should), we can just parse it for the path instead of having two locals here

annieetang replied:
asked sora and he said there isn't yet but I'll add it!

aatxe commented on tests/std/fs.test.luau:
```diff
@@ -133,6 +133,25 @@ test.suite("FsSuite", function(suite)
+		fs.createdirectory(nestedDir, { makeparents = true })
+		assert.eq(fs.exists(nestedDir), true)
+
+		-- TODO: create a fs.removedirectory with option to remove parents as well?
```
Not sure that this would really make sense (the remove parents option). How would you know when to stop? Do we just remove the whole disk?

annieetang replied:
haha yea that's why I didn't implement it yet cuz I realized that might not be a great option to have...

annieetang committed:
tmpdir in examples and make options optional

annieetang committed:
update setup cli commands
Merge branch 'primary' of https://github.com/luau-lang/lute into annietang/fs_createdirectory

aatxe commented on lute/std/libs/fs.luau:
```diff
@@ -15,6 +15,10 @@ export type watchhandle = fs.WatchHandle
 export type pathlike = pathlib.pathlike
 
+export type createdirectoryoptions = {
+	makeparents: boolean,
```
each individual option should probably be optional

annieetang committed:
make individual options in createdirectoryoptions optional
